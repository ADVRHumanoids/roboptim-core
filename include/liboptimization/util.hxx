// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Useful procedures (templated) implementation.
 */
#ifndef OPTIMIZATION_UTIL_HXX
# define OPTIMIZATION_UTIL_HXX
# include <boost/variant/apply_visitor.hpp>

# include <liboptimization/function.hh>

namespace optimization
{
  namespace detail
  {
    template <typename T>
    void
    jacobian_from_gradients (DerivableFunction::matrix_t& jac,
                             const std::vector<const T*>& c,
                             const DerivableFunction::vector_t& x)
    {
      for (unsigned i = 0; i < jac.size1 (); ++i)
        {
          DerivableFunction::jacobian_t grad = c[i]->jacobian (x);
          for (unsigned j = 0; j < jac.size2 (); ++j)
            jac (i, j) = grad(0, j);
        }
    }


    struct GenVariantPrintVisitor : public boost::static_visitor<void>
    {
      GenVariantPrintVisitor (std::ostream& o)
        : o_ (o)
      {}

      template <typename T>
      void operator () (const T& value)
      {
        o_ << value;
      }

      template <typename T>
      void operator () (const T* value)
      {
        if (!value)
          o_ << "Null pointer";
        else
          o_ << *value;
      }

    private:
      std::ostream& o_;
    };

  }; // end of namespace detail.

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream& o, const boost::variant<T1, T2>& var)
  {
    detail::GenVariantPrintVisitor pv (o);
    boost::apply_visitor (pv, var);
    return o;
  }

  template <typename T>
  std::ostream& operator<< (std::ostream& o, const std::vector<T>& vect)
  {
    typedef typename std::vector<T>::const_iterator citer_t;

    if (vect.empty ())
        return o << "Empty vector";

    citer_t it = vect.begin ();
    o << *it;
    ++it;

    for (; it != vect.end (); ++it)
      o << ", " << *it;
    return o;
  }

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream& o, const std::pair<T1, T2>& p)
  {
    return o << "(" << p.first << ", " << p.second << ")";
  }

}; // end of namespace optimization.

#endif //! OPTIMIZATION_UTIL_HXX
