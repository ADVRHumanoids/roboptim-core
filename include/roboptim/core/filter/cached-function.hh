// Copyright (C) 2010 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_FILTER_CACHED_FUNCTION_HH
# define ROBOPTIM_CORE_FILTER_CACHED_FUNCTION_HH
# include <roboptim/core/sys.hh>
# include <roboptim/core/debug.hh>

# include <map>

# include <boost/shared_ptr.hpp>

# include <roboptim/core/n-times-derivable-function.hh>

namespace roboptim
{
  /// \addtogroup roboptim_filter
  /// @{

  struct ltvector
  {
    bool operator()(const Function::vector_t& v1,
		    const Function::vector_t& v2) const
    {
      Function::vector_t::const_iterator it1 = v1.begin ();
      Function::vector_t::const_iterator it2 = v2.begin ();

      while (it1 != v1.end () && it2 != v2.end ())
	{
	  if (*it1 < *it2)
	    return true;
	  ++it1, ++it2;
	}
      return it1 != v1.end () && it2 == v2.end ();
    }
  };

  template <typename T>
  class ROBOPTIM_DLLAPI CachedFunction : public T
  {
  public:
    /// \brief Import value type.
    typedef typename DerivableFunction::value_type value_type;
    /// \brief Import size type.
    typedef typename DerivableFunction::size_type size_type;
    /// \brief Import vector type.
    typedef typename DerivableFunction::vector_t vector_t;
    /// \brief Import result type.
    typedef typename DerivableFunction::result_t result_t;
    /// \brief Import argument type.
    typedef typename DerivableFunction::argument_t argument_t;
    /// \brief Import gradient type.
    typedef typename DerivableFunction::gradient_t gradient_t;
    /// \brief Import hessian type.
    typedef typename TwiceDerivableFunction::hessian_t hessian_t;
    /// \brief Import jacobian type.
    typedef typename DerivableFunction::jacobian_t jacobian_t;
    /// \brief Import interval type.
    typedef typename DerivableFunction::interval_t interval_t;


    typedef std::map<Function::vector_t, Function::vector_t, ltvector>
      functionCache_t;

    explicit CachedFunction (boost::shared_ptr<const T> fct) throw ();
    ~CachedFunction () throw ();

    void reset () throw ();

  protected:
    virtual void impl_compute (result_t& result, const argument_t& argument)
      const throw ();


    virtual void impl_gradient (gradient_t& gradient,
				const argument_t& argument,
				size_type functionId = 0)
      const throw ();

    virtual void impl_hessian (hessian_t& hessian,
    			       const argument_t& argument,
    			       size_type functionId = 0) const throw ();

    virtual void impl_derivative (gradient_t& derivative,
    				  double argument,
    				  size_type order = 1) const throw ();

  private:
    boost::shared_ptr<const T> function_;
    mutable std::vector<functionCache_t> cache_;
    mutable std::vector<functionCache_t> gradientCache_;
    mutable std::vector<functionCache_t> hessianCache_;
  };

  /// @}

} // end of namespace roboptim

# include <roboptim/core/filter/cached-function.hxx>
#endif //! ROBOPTIM_CORE_FILTER_CACHED_FUNCTION_HH
