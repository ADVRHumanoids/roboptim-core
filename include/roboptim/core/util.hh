// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Useful procedures declaration.
 */
#ifndef ROBOPTIM_CORE_UTIL_HH
# define ROBOPTIM_CORE_UTIL_HH
# include <boost/variant.hpp>
# include <roboptim/core/derivable-function.hh>

namespace roboptim
{
  namespace detail
  {
    /// \internal
    /// Copy the content of a uBLAS vector into a C array.
    void vector_to_array (Function::value_type* dst, const Function::vector_t& src);

    /// \internal
    /// Copy the content of a C array into a uBLAS vector.
    void array_to_vector (Function::vector_t& dst, const Function::value_type* src);

    /// \internal
    /// Merge gradients from several functions (each gradient is a line).
    /// The first line of the jacobian is the only one used.
    template <typename T>
    void
    jacobian_from_gradients (DerivableFunction::matrix_t& jac,
                             const std::vector<const T*>& c,
                             const DerivableFunction::vector_t& x);
  } // end of namespace detail.

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream&, const boost::variant<T1, T2>&);

  template <typename T>
  std::ostream& operator<< (std::ostream&, const std::vector<T>&);

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream&, const std::pair<T1, T2>&);
} // end of namespace roboptim.

# include <roboptim/core/util.hxx>
#endif //! ROBOPTIM_CORE_UTIL_HH
