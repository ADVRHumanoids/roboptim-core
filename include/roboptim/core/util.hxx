// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_UTIL_HXX
# define ROBOPTIM_CORE_UTIL_HXX
# include <roboptim/core/function.hh>

namespace roboptim
{
  namespace detail
  {
    template <typename T>
    void
    jacobian_from_gradients (DerivableFunction::matrix_t& jac,
                             const std::vector<const T*>& c,
                             const DerivableFunction::vector_t& x)
    {
      for (unsigned i = 0; i < jac.size1 (); ++i)
        {
          DerivableFunction::jacobian_t grad = c[i]->jacobian (x);
          for (unsigned j = 0; j < jac.size2 (); ++j)
            jac (i, j) = grad(0, j);
        }
    }
  } // end of namespace detail.

  template <typename T>
  std::ostream& operator<< (std::ostream& o, const std::vector<T>& vect)
  {
    typedef typename std::vector<T>::const_iterator citer_t;

    if (vect.empty ())
        return o << "Empty vector";

    citer_t it = vect.begin ();
    o << *it;
    ++it;

    for (; it != vect.end (); ++it)
      o << ", " << *it;
    return o;
  }

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream& o, const std::pair<T1, T2>& p)
  {
    return o << "(" << p.first << ", " << p.second << ")";
  }

}; // end of namespace roboptim.

#endif //! ROBOPTIM_CORE_UTIL_HXX
