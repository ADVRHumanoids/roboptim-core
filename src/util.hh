// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \file src/util.hh
 *
 * \brief Useful procedures declaration.
 */
#ifndef OPTIMIZATION_UTIL_HH
# define OPTIMIZATION_UTIL_HH
# include <function.hh>
# include <problem.hh>

namespace optimization
{
  namespace detail
  {
    /// Copy the content of a uBLAS vector into a C array.
    void vector_to_array (Function::value_type* dst, const Function::vector_t& src);

    /// Copy the content of a C array into a uBLAS vector.
    void array_to_vector (Function::vector_t& dst, const Function::value_type* src);

    /// Merge gradients from several functions (each gradient is a line).
    void jacobian_from_gradients (Function::matrix_t&,
                                  const Problem::constraints_t&,
                                  const Function::vector_t&);

  }; // end of namespace detail.
}; // end of namespace optimization.

#endif //! OPTIMIZATION_UTIL_HH
