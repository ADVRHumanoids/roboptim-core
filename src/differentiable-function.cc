// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "debug.hh"

#include "roboptim/core/differentiable-function.hh"
#include "roboptim/core/indent.hh"
#include "roboptim/core/util.hh"

namespace roboptim
{
  DifferentiableFunction::DifferentiableFunction (size_type inputSize,
						  size_type outputSize,
						  std::string name)
    throw ()
    : Function (inputSize, outputSize, name)
  {
  }

  void
  DifferentiableFunction::impl_jacobian (jacobian_t& jacobian,
					 const argument_t& argument)
    const throw ()
  {
    for (unsigned i = 0; i < outputSize (); ++i)
      {
        gradient_t grad = gradient (argument, i);
        for (unsigned j = 0; j < inputSize (); ++j)
          jacobian (i, j) = grad[j];
      }
  }

  std::ostream&
  DifferentiableFunction::print (std::ostream& o) const throw ()
  {
    if (getName ().empty ())
      return o << "Differentiable function";
    else
      return o << getName () << " (derivable function)";
  }

} // end of namespace roboptim
