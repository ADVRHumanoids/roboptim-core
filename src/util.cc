// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "debug.hh"

#include <cstring>
#include <boost/numeric/ublas/matrix.hpp>

#include "roboptim/core/util.hh"

namespace roboptim
{
  namespace detail
  {
    namespace ublas = boost::numeric::ublas;

    void
    vector_to_array (Function::value_type* dst, const Function::vector_t& src)
    {
      if (src.empty ())
	return;
      memcpy (dst, &src[0], src.size () * sizeof (Function::value_type));

      // NaN != NaN, handle this case.
      for (std::size_t i = 0; i < src.size (); ++i)
	if (src[i] != src[i])
	  assert (dst[i] != dst[i]);
	else
	  assert (dst[i] == src[i]);
    }

    void
    array_to_vector (Function::vector_t& dst, const Function::value_type* src)
    {
      if (dst.empty ())
	return;
      memcpy (&dst[0], src, dst.size () * sizeof (Function::value_type));

      // NaN != NaN, handle this case.
      for (std::size_t i = 0; i < dst.size (); ++i)
	if (src[i] != src[i])
	  assert (dst[i] != dst[i]);
	else
	  assert (dst[i] == src[i]);
    }
  } // end of namespace detail.
} // end of namespace roboptim.
