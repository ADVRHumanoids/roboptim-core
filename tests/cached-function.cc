// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/common.hh"

#include <iostream>

#include <roboptim/core/io.hh>
#include <roboptim/core/differentiable-function.hh>
#include <roboptim/core/util.hh>
#include <roboptim/core/filter/cached-function.hh>

using namespace roboptim;

boost::shared_ptr<boost::test_tools::output_test_stream> output;

struct F : public DifferentiableFunction
{
  F () : DifferentiableFunction (1, 1, "2 * x")
  {}

  void impl_compute (result_t& res, const argument_t& argument) const throw ()
  {
    (*output) << "computation (not cached)" << std::endl;
    res.setZero ();
    res[0] = 2. * argument[0];
  }

  void impl_gradient (gradient_t& grad, const argument_t&,
		      size_type) const throw ()
  {
    (*output) << "gradient computation (not cached)" << std::endl;
    grad.setZero ();
    grad[0] = 2.;
  }
};

BOOST_AUTO_TEST_CASE (cached_function)
{
  output = retrievePattern ("cached-function");

  boost::shared_ptr<F> f (new F ());

  CachedFunction<DifferentiableFunction> cachedF (f);

  (*output) << cachedF << ":" << std::endl
	    << std::endl;

  Function::vector_t x (1);
  for (double i = 0.; i < 10.; i += 0.5)
  {
    x[0] = i;
    (*output) << cachedF (x) << std::endl;
    (*output) << cachedF (x) << std::endl;
    BOOST_CHECK_EQUAL ((*f) (x)[0], cachedF (x)[0]);

    (*output) << cachedF.gradient (x) << std::endl;
    (*output) << cachedF.gradient (x) << std::endl;
  }

  std::cout << output->str () << std::endl;
  BOOST_CHECK (output->match_pattern ());
}
