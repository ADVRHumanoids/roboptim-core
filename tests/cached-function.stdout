2 * x (cached) (derivable function):

not cached
[1](0)
cached
[1](0)
cached
[1](2)
[1](2)
not cached
[1](1)
cached
[1](1)
cached
[1](2)
[1](2)
not cached
[1](2)
cached
[1](2)
cached
[1](2)
[1](2)
not cached
[1](3)
cached
[1](3)
cached
[1](2)
[1](2)
not cached
[1](4)
cached
[1](4)
cached
[1](2)
[1](2)
not cached
[1](5)
cached
[1](5)
cached
[1](2)
[1](2)
not cached
[1](6)
cached
[1](6)
cached
[1](2)
[1](2)
not cached
[1](7)
cached
[1](7)
cached
[1](2)
[1](2)
not cached
[1](8)
cached
[1](8)
cached
[1](2)
[1](2)
not cached
[1](9)
cached
[1](9)
cached
[1](2)
[1](2)
not cached
[1](10)
cached
[1](10)
cached
[1](2)
[1](2)
not cached
[1](11)
cached
[1](11)
cached
[1](2)
[1](2)
not cached
[1](12)
cached
[1](12)
cached
[1](2)
[1](2)
not cached
[1](13)
cached
[1](13)
cached
[1](2)
[1](2)
not cached
[1](14)
cached
[1](14)
cached
[1](2)
[1](2)
not cached
[1](15)
cached
[1](15)
cached
[1](2)
[1](2)
not cached
[1](16)
cached
[1](16)
cached
[1](2)
[1](2)
not cached
[1](17)
cached
[1](17)
cached
[1](2)
[1](2)
not cached
[1](18)
cached
[1](18)
cached
[1](2)
[1](2)
not cached
[1](19)
cached
[1](19)
cached
[1](2)
[1](2)
